From b8646639be19e99529d95ed5b83278d4f49fca8c Mon Sep 17 00:00:00 2001
From: Hugues Fruchet <hugues.fruchet@foss.st.com>
Date: Thu, 11 Dec 2025 17:19:17 +0100
Subject: [PATCH] media: verisilicon: fix decoding of jpeg with restart
 interval

Fix decoding of JPEG having restart interval markers.

Change-Id: I130186cd38400c33ad40d3cc50d4deadcce433ab
Signed-off-by: Hugues Fruchet <hugues.fruchet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/519342
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Hugues FRUCHET <hugues.fruchet@st.com>
Tested-by: Hugues FRUCHET <hugues.fruchet@st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
---
 drivers/media/platform/verisilicon/hantro_g1_jpeg_dec.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

--- a/drivers/media/platform/verisilicon/hantro_g1_jpeg_dec.c
+++ b/drivers/media/platform/verisilicon/hantro_g1_jpeg_dec.c
@@ -390,8 +390,12 @@ int hantro_g1_jpeg_dec_run(struct hantro
 
 	reg |= G1_REG_DEC_CTRL2_STRM_START_BIT(stream_bit_offset(&header));
 
-	if (header.restart_interval)
+	if (header.restart_interval) {
 		reg |= G1_REG_DEC_CTRL2_SYNC_MARKER_E;
+		vdpu_write_relaxed(vpu,
+				   G1_REG_DEC_CTRL5_PJPEG_REST_FREQ(header.restart_interval),
+				   G1_REG_DEC_CTRL5);
+	}
 
 	vdpu_write_relaxed(vpu, reg, G1_REG_DEC_CTRL2);
 
